### PR TITLE
provider/gce: VPN resources, documentation, tests and example

### PR DESCRIPTION
**Pull request by lwander**
These were the missing components required for creating VPNs in GCE using Terraform.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ FEATURES:
   * **New resource: `google_compute_project_metadata`** [GH-3065]
   * **New resources: `aws_ami`, `aws_ami_copy`, `aws_ami_from_instance`** [GH-2874]
   * **New resource: `google_storage_bucket_object`** [GH-3192]
+  * **New resources: `google_compute_vpn_gateway`, `google_compute_vpn_tunnel`** [GH-3213]
```
Comment (radeksimko) [inline CHANGELOG.md]: FYI: The changelog is something maintainers update via github web iface when they're merging the PR to avoid any potential conflicts.

File: builtin/providers/google/resource_compute_address.go [hunk 2]
```diff
@@ -13,37 +13,51 @@ import (
...
+			"region": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
 		},
 	}
 }
 
+func getOptionalRegion(d *schema.ResourceData, config *Config) string {
```
Comment (radeksimko) [inline builtin/providers/google/resource_compute_address.go]: I wish we could just use this as `DefaultFunc`, but because of that dependency (`*Config`), it is effectively impossible. :-/

File: builtin/providers/google/resource_compute_vpn_gateway.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	description := d.Get("description").(string)
+	region := d.Get("region").(string)
+	project := config.Project
+
+	vpnGatewaysService := compute.NewTargetVpnGatewaysService(config.clientCompute)
+
+	vpnGateway := &compute.TargetVpnGateway{
+		Name:		 name,
+		Network:	 network,
+		Description: description,
```
Comment (radeksimko) [inline builtin/providers/google/resource_compute_vpn_gateway.go]: `gofmt` would probably help here...

File: CHANGELOG.md (hunk 1 above)
Comment (lwander) [inline CHANGELOG.md]: Ah, I was hoping I was saving you the trouble of updating that, sorry!

File: builtin/providers/google/resource_compute_vpn_gateway.go (hunk 3 above)
Comment (lwander) [inline builtin/providers/google/resource_compute_vpn_gateway.go]: That seems really useful, thank you!

File: builtin/providers/google/resource_compute_address.go [hunk 4]
```diff
@@ -114,10 +131,10 @@ func resourceComputeAddressDelete(d *schema.ResourceData, meta interface{}) erro
 	// Wait for the operation to complete
 	w := &OperationWaiter{
 		Service: config.clientCompute,
-		Op:      op,
+		Op:		 op,
```
Comment (radeksimko) [inline builtin/providers/google/resource_compute_address.go]: The original formatting is `gofmt`-compliant - i.e. spaces were ok here, no need to change this to tabs.

File: builtin/providers/google/resource_compute_vpn_tunnel.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+	if ikeVersion < 1 || ikeVersion > 2 {
+		return fmt.Errorf("Only IKE version 1 or 2 supported, not %d", ikeVersion)
+	}
+
+	vpnTunnelsService := compute.NewVpnTunnelsService(config.clientCompute)
+
+	vpnTunnel := &compute.VpnTunnel{
+		Name:			  name,
+		Description:	  description,
```
Comment (radeksimko) [inline builtin/providers/google/resource_compute_vpn_tunnel.go]: Description and a few other fields are optional, but it's always being passed to the API - do you merely rely on the marshalling logic of SDK and/or API that will ignore this?

The convention in most other resources is to use `d.GetOk()` for optional fields, so these are only touched if defined by the user.

File: builtin/providers/google/resource_compute_vpn_gateway.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	}
+}
+
+func resourceComputeVpnGatewayCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	name := d.Get("name").(string)
+	network := d.Get("network").(string)
+	description := d.Get("description").(string)
+	region := d.Get("region").(string)
```
Comment (radeksimko) [inline builtin/providers/google/resource_compute_vpn_gateway.go]: If you started using `getOptionalRegion` above, why not use it in these two new resources as well?
Even though it doesn't feel as elegant as if we were able to use it directly in the schema via `DefaultFunc`, I think it's a nice feature and almost expected behaviour - less required fields that can be inherited from the provider.

File: website/source/docs/providers/google/r/compute_vpn_gateway.html.markdown [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    depends_on = ["google_compute_forwarding_rule.fr_esp",
+        "google_compute_forwarding_rule.fr_udp500",
+        "google_compute_forwarding_rule.fr_udp4500"]
+}
+
+resource "google_compute_route" "route1" {
+    name = "route1"
+    network = "${google_compute_network.network1.name}"
+    next_hop_vpn_tunnel = "${google_compute_vpn_tunnel.tunnel1.self_link}"
+    dest_range = NETWORK_RANGE_OF_OTHER_NETWORK
```
Comment (radeksimko) [inline website/source/docs/providers/google/r/compute_vpn_gateway.html.markdown]: I think it would be better to replace this with an example - so it's obvious that CIDR is expected.

File: website/source/docs/providers/google/r/compute_vpn_tunnel.html.markdown [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    depends_on = ["google_compute_forwarding_rule.fr_esp",
+        "google_compute_forwarding_rule.fr_udp500",
+        "google_compute_forwarding_rule.fr_udp4500"]
+}
+
+resource "google_compute_route" "route1" {
+    name = "route1"
+    network = "${google_compute_network.network1.name}"
+    next_hop_vpn_tunnel = "${google_compute_vpn_tunnel.tunnel1.self_link}"
+    dest_range = NETWORK_RANGE_OF_OTHER_NETWORK
```
Comment (radeksimko) [inline website/source/docs/providers/google/r/compute_vpn_tunnel.html.markdown]: as above.

File: examples/gce-vpn/variables.tf [hunk 9]
```diff
@@ -0,0 +1,7 @@
+variable "project" {
```
Comment (radeksimko) [inline examples/gce-vpn/variables.tf]: :+1: for example! This will help users to get up to speed, it would be even more helpful if you add a brief Readme describing the setup, similar to the ones we already have for each example.
https://github.com/hashicorp/terraform/tree/master/examples

**Comment by radeksimko** [comment]
I ran both acceptance tests - worked aok. Just left you a few notes about docs, formatting and schema.
Otherwise LGTM.

File: builtin/providers/google/resource_compute_vpn_tunnel.go (hunk 5 above)
Comment (lwander) [inline builtin/providers/google/resource_compute_vpn_tunnel.go]: You're right, it's definitely neater to user `GetOk()` here.

File: website/source/docs/providers/google/r/compute_vpn_gateway.html.markdown (hunk 7 above)
Comment (lwander) [inline website/source/docs/providers/google/r/compute_vpn_gateway.html.markdown]: Do you mean hardcode a CIDR address?

File: builtin/providers/google/resource_compute_vpn_gateway.go (hunk 6 above)
Comment (lwander) [inline builtin/providers/google/resource_compute_vpn_gateway.go]: I'll add it in here, and go through the other resources that use `region` as well for a separate PR if you think that's best.

File: website/source/docs/providers/google/r/compute_vpn_gateway.html.markdown (hunk 7 above)
Comment (radeksimko) [inline website/source/docs/providers/google/r/compute_vpn_gateway.html.markdown]: yes, pretty much like in the example of `google_compute_route` docs:
https://www.terraform.io/docs/providers/google/r/compute_route.html
`dest_range = "15.0.0.0/24"`

File: builtin/providers/google/resource_compute_vpn_gateway.go (hunk 6 above)
Comment (radeksimko) [inline builtin/providers/google/resource_compute_vpn_gateway.go]: yep, I think that's best.

If you'll come up with anything more elegant than calling a function in each of these CRUD functions, it will be even better, but if not, the current solution is :ok_hand: .

Due to the nature of schema (which is typically validated before any API calls are made) it would probably require much bigger changes in the core.

File: website/source/docs/providers/google/r/compute_vpn_gateway.html.markdown [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    ip_protocol = "UDP"
+    port_range = "4500"
+    ip_address = "${google_compute_address.vpn_static_ip.address}"
+    target = "${google_compute_vpn_gateway.target_gateway.self_link}"
+}
+
+resource "google_compute_vpn_tunnel" "tunnel1" {
+    name = "tunnel1"
+    region = "${var.region}"
+    peer_ip = 15.0.0.120
```
Comment (radeksimko) [inline website/source/docs/providers/google/r/compute_vpn_gateway.html.markdown]: Can you also escape this as string? otherwise it's invalid syntax. :wink:

File: CHANGELOG.md (hunk 1 above)
Comment (sparkprime) [inline CHANGELOG.md]: I didn't know you could do it when merging, I used to do a separate commit after the merge.  How do you do it?

Comment (radeksimko) [inline CHANGELOG.md]: I usually just do it via github web iface. AFAIK others do as well.

**Comment by lwander** [comment]
I squashed my changes, and modified (and tested) the example to set up a VPN between networks in different regions. @sparkprime

**Comment by sparkprime** [comment]
LGTM
